lab4/lab4/BalancedTreeSearch.cpp

```diff
@@ -0,0 +1,211 @@
+#include <iostream>
+#include "BalancedTreeSearch.h"
+
+BalancedTreeSearch::BalancedTreeSearch(const BalancedTreeSearch& other) {
+    m_root = copyTree(other.m_root);
+}
+
+BalancedTreeSearch::BalancedTreeSearch(BalancedTreeSearch&& other) noexcept {
+    m_root = other.m_root;
+    other.m_root = nullptr;
+}
+
+int BalancedTreeSearch::heightDifference(const Node* node) const {
+    if (!node) {
+        return 0;
+    }
+
+    int leftHeight = height(node->leftChild());
+    int rightHeight = height(node->rightChild());
+
+    return rightHeight - leftHeight;
+}
+
+void BalancedTreeSearch::newHeight(BalancedTreeSearch::Node* node) {
+    if (node) {
+        int leftHeight = height(node->leftChild());
+        int rightHeight = height(node->rightChild());
+        node->setHeight(std::max(leftHeight, rightHeight) + 1);
+    }
+}
+
+BalancedTreeSearch::Node* BalancedTreeSearch::balancing(Node* node, bool& isFixed) {
```

Comment:
Как вариант, добавить флаг, балансируем после добавления или удаления, чтобы можно было скорректировать условия прекращения балансировки.
```suggestion
BalancedTreeSearch::Node* BalancedTreeSearch::balancing(Node* node, bool& isFixed) {
```